llvm/lib/Transforms/Vectorize/VPlanConstantFolder.h

```diff
@@ -0,0 +1,91 @@
+//===- VPlanConstantFolder.h - ConstantFolder for VPlan -------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "VPlanValue.h"
+#include "llvm/IR/ConstantFolder.h"
+
+namespace llvm {
+class VPConstantFolder {
+private:
+  ConstantFolder Folder;
+
+  Constant *getIRConstant(VPValue *V) const {
+    return dyn_cast_if_present<Constant>(V->getUnderlyingValue());
```

Comment:
Using the underlying IR value is only safe for live-in VPValues; for recipes, the underlying value should not have any semantic meaning (not 100% true yet for all recipes, but we are getting there)

Reply:
Oh, I saw that this wasn't true in practice, and even put up https://github.com/llvm/llvm-project/pull/128886, which includes some improvements. I do have a question though: isn't the defining recipe created with several intermediate Builder-create calls? Why should the defining recipe be present when we're in the process of building recipes?